File: 98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        TreeNode *curr = root, *prev = NULL, *pred = NULL;
        bool valid = true;
        while (curr != NULL) {
            if (curr->left == NULL) {
                if (prev != NULL) {
                    if ((curr->val) <= (prev->val)) {
                        valid = false;
                    }
                }
                prev = curr;
                curr = curr->right;
            } else {
                pred = curr->left;
                while (pred->right != NULL && pred->right != curr) {
                    pred = pred->right;
                }
                if (pred->right == curr) {
                    pred->right = NULL;
                    if (prev != NULL) {
                        if (curr->val <= prev->val) {
                            valid = false;
                        }
                    }
                    prev = curr;
                    curr = curr->right;
                } else {
                    pred->right = curr;
                    curr = curr->left;
                }
            }
        }
        return valid;
    }
};

File: java/98. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// O(n)
class Solution {
    public boolean isValidBST(TreeNode root) {
        TreeNode pre = null;
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cur = root;
        while (cur != null || !stack.isEmpty()) {
            if (cur == null) {
                cur = stack.pop();
                if (pre != null && pre.val >= cur.val) {
                    return false;
                }
                pre = cur;
                cur = cur.right;
            } else {
                stack.push(cur);
                cur = cur.left;
            }
        }
        return true;
    }
}
